Bottom: fdd492cdb13406bafcf14c37b554b2de76a99d5b
Top:    a70e486d7fb3d18adace33aea325353d847f6b41
Author: Sean Hefty <sean.hefty@intel.com>
Date:   2012-08-07 10:30:05 -0700

Refresh of fstat

---

diff --git a/src/preload.c b/src/preload.c
index 085e8b4..b9c32e6 100644
--- a/src/preload.c
+++ b/src/preload.c
@@ -84,6 +84,7 @@ struct socket_calls {
 			  void *optval, socklen_t *optlen);
 	int (*fcntl)(int socket, int cmd, ... /* arg */);
 	int (*dup2)(int oldfd, int newfd);
+	int (*fstat)(int fd, struct stat *buf);
 };
 
 static struct socket_calls real;
@@ -259,6 +260,7 @@ static void init_preload(void)
 	real.getsockopt = dlsym(RTLD_NEXT, "getsockopt");
 	real.fcntl = dlsym(RTLD_NEXT, "fcntl");
 	real.dup2 = dlsym(RTLD_NEXT, "dup2");
+	real.fstat = dlsym(RTLD_NEXT, "fstat");
 
 	rs.socket = dlsym(RTLD_DEFAULT, "rsocket");
 	rs.bind = dlsym(RTLD_DEFAULT, "rbind");
@@ -960,3 +962,17 @@ int dup2(int oldfd, int newfd)
 	atomic_inc(&oldfdi->refcnt);
 	return newfd;
 }
+
+int fstat(int socket, struct stat *buf)
+{
+	int fd, ret;
+
+	if (fd_get(socket, &fd) == fd_rsocket) {
+		ret = real.fstat(socket, buf);
+		if (!ret)
+			buf->st_mode |= __S_IFSOCK;
+	} else {
+		ret = real.fstat(fd, buf);
+	}
+	return ret;
+}
